riscv-pac/macros/src/lib.rs

```diff
@@ -0,0 +1,367 @@
+            Data::Enum(data) => &data.variants,
+            _ => panic!("Input is not an enum"),
+        };
+        for v in variants.iter() {
+            let ident = v.ident.clone();
+            let value = match &v.discriminant {
+                Some(d) => match &d.1 {
+                    syn::Expr::Lit(expr_lit) => match &expr_lit.lit {
+                        syn::Lit::Int(lit_int) => match lit_int.base10_parse::<usize>() {
+                            Ok(num) => num,
+                            Err(_) => {
+                                panic!("All variant discriminants must be unsigned integers")
+                            }
+                        },
+                        _ => panic!("All variant discriminants must be unsigned integers"),
+                    },
+                    _ => panic!("All variant discriminants must be unsigned integers"),
+                },
+                _ => panic!("Variant must have a discriminant"),
+            };
+
+            if numbers.insert(value, ident).is_some() {
+                panic!("Duplicate discriminant value");
+            }
+            if value > max_number {
+                max_number = value;
+            }
+        }
+
+        Self {
+            name,
+            max_number,
+            numbers,
+        }
+    }
+
+    /// Returns a token stream representing the maximum discriminant value of the enum
+    fn max_discriminant(&self) -> TokenStream2 {
+        TokenStream2::from_str(&format!("{}", self.max_number)).unwrap()
+    }
+
+    /// Returns a vector of token streams representing the valid matches in the `pac::from_number` function
+    fn valid_matches(&self) -> Vec<TokenStream2> {
+        self.numbers
+            .iter()
+            .map(|(num, ident)| {
+                TokenStream2::from_str(&format!("{num} => Ok(Self::{ident})")).unwrap()
+            })
+            .collect()
+    }
+
+    /// Returns a vector of token streams representing the interrupt handler functions
+    fn interrupt_handlers(&self) -> Vec<TokenStream2> {
+        self.numbers
+            .values()
+            .map(|ident| {
+                quote! { fn #ident () }
+            })
+            .collect()
+    }
+
+    /// Returns a sorted vector of token streams representing all the elements of the interrupt array.
+    /// If an interrupt number is not present in the enum, the corresponding element is `None`.
+    /// Otherwise, it is `Some(<interrupt_handler>)`.
+    fn interrupt_array(&self) -> Vec<TokenStream2> {
+        let mut vectors = vec![];
+        for i in 0..=self.max_number {
+            if let Some(ident) = self.numbers.get(&i) {
+                vectors.push(quote! { Some(#ident) });
+            } else {
+                vectors.push(quote! { None });
+            }
+        }
+        vectors
+    }
+
+    fn vector_table(&self) -> TokenStream2 {
+        let mut asm = String::from(
+            r#"
+core::arch::global_asm!("
```

Comment:
```suggestion
#[cfg(any(target_arch = "riscv32", target_arch = "riscv64"))]
core::arch::global_asm!("
```

Fixes the `ui` tests.